### PR TITLE
Add optional `force` argument to `refresh_continuous_aggregate`

### PR DESCRIPTION
**Pull request by zilder**
Following the [SDC](https://github.com/timescale/Support-Dev-Collab/issues/2331). Continuous aggregates that are built on top of tiered hypertables cannot access and materialize tiered data when the `timescaledb.enable_tiered_reads` GUC is disabled at the server level. And there is no way for a user to manually force the refresh with tiered reads enabled. Here we add an optional `force` parameter to the `refresh_continuous_aggregate` procedure that would allow user to partially re-materialize cagg within a time window that has already been materialized.

### CODE REVIEW

File: tsl/test/expected/cagg_refresh.out [hunk 1]
```diff
@@ -489,3 +489,41 @@ SELECT time_bucket('7 days', time) AS day, device, avg(temp) AS avg_temp
...
+ Mon May 04 00:00:00 2020 UTC |      0 | 15.7647058823529
+ Mon May 04 00:00:00 2020 UTC |      1 | 24.3142857142857
+ Mon May 04 00:00:00 2020 UTC |      2 | 14.8205128205128
+ Mon May 04 00:00:00 2020 UTC |      3 | 18.1111111111111
+ Tue May 05 00:00:00 2020 UTC |      0 | 19.3846153846154
+ Tue May 05 00:00:00 2020 UTC |      1 | 16.5555555555556
+ Tue May 05 00:00:00 2020 UTC |      2 | 18.5714285714286
+ Tue May 05 00:00:00 2020 UTC |      3 | 23.5714285714286
+(8 rows)
+
```
Comment (fabriziomello) [inline tsl/test/expected/cagg_refresh.out:529]: Would be good to move those tests to `include/cagg_refresh_common.sql` so we will be able to test using the current materialization code (DELETE + INSERT) and the new one (MERGE + DELETE)

**Comment by erimatnor** [comment]
I think we discussed this previously, and one idea was to invalidate the region covered by the tiered data when the GUC to enable tiered refreshing was changed to "true/on". This would not require changes to our API and is the natural way to make the region refresh with current mechanisms. Was that approach discarded for some reason? If so, can someone explain what disqualified it?

**Comment by zilder** [comment]
I probably didn't catch that. Do you mean that in the GUC callback we set some sort of a flag that would indicate that we should insert an invalidation record whenever we refresh a range? I see a couple of issues:
1. Customer decides to change the GUC on server level. Now if they want to refresh missing aggregates over tiered data they would need to turn the GUC "off" and then "on" again.
2. This would alter the behavior of the refresh procedure for regular caggs (over non-tiered data).
In both cases it feels like not a clear UX.

**Comment by fabriziomello** [comment]
> I think we discussed this previously, and one idea was to invalidate the region covered by the tiered data when the GUC to enable tiered refreshing was changed to "true/on". This would not require changes to our API and is the natural way to make the region refresh with current mechanisms. Was that approach discarded for some reason? If so, can someone explain what disqualified it?

The problem with this approach is the case when the tiered data is already materialized that will generate unnecessary invalidation logs.

**Comment by erimatnor** [comment]
> > I think we discussed this previously, and one idea was to invalidate the region covered by the tiered data when the GUC to enable tiered refreshing was changed to "true/on". This would not require changes to our API and is the natural way to make the region refresh with current mechanisms. Was that approach discarded for some reason? If so, can someone explain what disqualified it?
> 
> The problem with this approach is the case when the tiered data is already materialized that will generate unnecessary invalidation logs.


> @erimatnor
> 
> > one idea was to invalidate the region covered by the tiered data when the GUC to enable tiered refreshing was changed to "true/on"
> 
> I probably didn't catch that. Do you mean that in the GUC callback we set some sort of a flag that would indicate that we should insert an invalidation record whenever we refresh a range? I see a couple of issues:
> 

I meant adding the invalidations in the GUC, not setting a flag. Nothing else needs to change.

>     1. Customer decides to change the GUC on server level. Now if they want to refresh missing aggregates over tiered data they would need to turn the GUC "off" and then "on" again.

Not sure I follow why they'd have to torn off and on again. I believe guc callbacks are called on parsing the server config file as well.

> 
>     2. This would alter the behavior of the refresh procedure for regular caggs (over non-tiered data).
> 

I don't think this would be necessary if the invalidations are added via the guc callback.

> 
> In both cases it feels like not a clear UX.

The idea is that this would be completely transparent and no need for user to learn about a `force` parameter, so IMO the UX is better because it "just works" as expected.

With the other approach, the user first has to enable the GUC for tiered caggs to work, then use the `force` parameter for the first refresh. That seems like a strictly worse UX to me.

**Comment by erimatnor** [comment]
> > I think we discussed this previously, and one idea was to invalidate the region covered by the tiered data when the GUC to enable tiered refreshing was changed to "true/on". This would not require changes to our API and is the natural way to make the region refresh with current mechanisms. Was that approach discarded for some reason? If so, can someone explain what disqualified it?
> 
> The problem with this approach is the case when the tiered data is already materialized that will generate unnecessary invalidation logs.

That's just one large invalidation entry per affected cagg. It should not be an issue right?

**Comment by fabriziomello** [comment]
> > > I think we discussed this previously, and one idea was to invalidate the region covered by the tiered data when the GUC to enable tiered refreshing was changed to "true/on". This would not require changes to our API and is the natural way to make the region refresh with current mechanisms. Was that approach discarded for some reason? If so, can someone explain what disqualified it?
> > 
> > 
> > The problem with this approach is the case when the tiered data is already materialized that will generate unnecessary invalidation logs.
> 
> That's just one large invalidation entry per affected cagg. It should not be an issue right? 

It is a problem (actually this is exact the problem I want to avoid) because the current architecture don't deal with large ranges very well and doing that will make things worse for the customer.

> Also, can't we detect whether it is up-to-date or not and avoid adding the invalidation? If we can't detect it, it would be "correct" to add the invalidation AFAICT.

We can check if exists some tiered data already materialized but it don't guarantee that everything is actually materialized and this is the reason to have this `force` to make sure we'll rematerialize what is needed.

**Comment by erimatnor** [comment]
> > > > I think we discussed this previously, and one idea was to invalidate the region covered by the tiered data when the GUC to enable tiered refreshing was changed to "true/on". This would not require changes to our API and is the natural way to make the region refresh with current mechanisms. Was that approach discarded for some reason? If so, can someone explain what disqualified it?
> > > 
> > > 
> > > The problem with this approach is the case when the tiered data is already materialized that will generate unnecessary invalidation logs.
> > 
> > 
> > That's just one large invalidation entry per affected cagg. It should not be an issue right?
> 
> It is a problem (actually this is exact the problem I want to avoid) because the current architecture don't deal with large ranges very well and doing that will make things worse for the customer.
> 
Ok, sounds like a bigger problem then. It is just that my alarm bells start to ring when I see that we are trying to build around the current architecture. I'd prefer we fix the underlying problem, but maybe this is not possible in the short term.

> 
> We can check if exists some tiered data already materialized but it don't guarantee that everything is actually materialized and this is the reason to have this `force` to make sure we'll rematerialize what is needed.

But it will only work for one cagg, so you'd have to manually force refresh  all caggs that are connected to a tiered hypertable. If you forget one or more, you are now actually in a bad state where some caggs aren't accurately representing the data even if you refresh normally.  In contrast, adding the invalidation will add the proper state to remember that this region is no longer valid and needs a refresh. This is what invalidations are for. But now we are saying we cannot use invalidations because they don't work well. So, now we use them sometimes, but in other times we don't. To me that sounds like a system where you cannot fully trust that a refresh is actually a refresh, which might lead to users _always_ forcing a refresh just to be sure.

Anyway, I am just pointing these things out to make sure we thought this through, but if you are convinced that relying on the user forcing a refresh is the best option right now, then I trust you that there is no better option.

File: tsl/test/expected/cagg_refresh.out (hunk 1 above)
Comment (zilder) [reply tsl/test/expected/cagg_refresh.out]: I moved the test to `chunk_utils_internal.sql` to test it with fdw chunk. Do you think it is necessary to add a test to `cagg_refresh_common`? I haven't touched the materialization code, only the part where the invalidation records combine.

Comment (fabriziomello) [reply tsl/test/expected/cagg_refresh.out]: Would be good to have a test on `cagg_refresh_common` because it will be exerciced more and by two materialization strategies: (DELETE + INSERT) and (MERGE + DELETE).

Comment (fabriziomello) [reply tsl/test/expected/cagg_refresh.out]: IMHO this test is enough if added to `cagg_refresh_common`

File: tsl/test/sql/include/cagg_refresh_common.sql [hunk 2]
```diff
@@ -106,6 +106,31 @@ CALL refresh_continuous_aggregate('daily_temp', '2020-05-01'::text, '2020-05-03'
...
+-- the materialization hypertable and forefully re-materialize it.
+SELECT ht.schema_name || '.' || ht.table_name AS mat_ht, mat_hypertable_id FROM _timescaledb_catalog.continuous_agg cagg
+JOIN _timescaledb_catalog.hypertable ht ON cagg.mat_hypertable_id = ht.id
+WHERE user_view_name = 'daily_temp' \gset
+
+-- Delete the data from the materialization hypertable
+DELETE FROM :mat_ht;
+
+-- Run regular refresh, it should not touch previously materialized range
+-- CALL refresh_continuous_aggregate('daily_temp', '2020-05-04 00:00 UTC', '2020-05-05 00:00 UTC');
```
Comment (fabriziomello) [inline tsl/test/sql/include/cagg_refresh_common.sql:124]: Leftover?

**Review by fabriziomello — APPROVED** [review]

**Review by gayyappan — APPROVED** [review]

File: tsl/test/expected/cagg_refresh.out [hunk 3]
```diff
@@ -223,6 +223,47 @@ psql:include/cagg_refresh_common.sql:105: ERROR:  invalid time argument type "te
...
 psql:include/cagg_refresh_common.sql:106: ERROR:  invalid time argument type "integer"
 \set ON_ERROR_STOP 1
+-- Test forceful refreshment. Here we simulate the situation that we've seen
+-- with tiered data when `timescaledb.enable_tiered_reads` were disabled on the
+-- server level. In that case we would not see materialized tiered data and
+-- we wouldn't be able to re-materialize the data using a normal refresh call
+-- because it would skip previously materialized ranges, but it should be
+-- possible with `force=>true` parameter. To simulate this use-case we clear
+-- the materialization hypertable and forefully re-materialize it.
+SELECT ht.schema_name || '.' || ht.table_name AS mat_ht, mat_hypertable_id FROM _timescaledb_catalog.continuous_agg cagg
```
Comment (mkindahl) [inline tsl/test/expected/cagg_refresh.out:233]: Nit: If the schema name or table name is "strange", this code might generate something that does not parse as a table. Compare these:

```sql
mats=# select format('%I.%I', nspname, relname), nspname || '.' || relname from pg_class cls join pg_namespace nsp on relnamespace = nsp.oid where nspname = 'public';
                format                 |               ?column?                
---------------------------------------+---------------------------------------
 public."fnupp'"                       | public.fnupp'
 public.foo                            | public.foo
 public.conditions                     | public.conditions
 public.conditions_time_idx            | public.conditions_time_idx
 public.events_event_id_seq            | public.events_event_id_seq
 public.events_updated_at_event_id_idx | public.events_updated_at_event_id_idx
 public.events                         | public.events
 public.unique_key3                    | public.unique_key3
 public.testtable3_time_idx3           | public.testtable3_time_idx3
 public.testtable3                     | public.testtable3
 public.metric                         | public.metric
(11 rows)
```


```suggestion
SELECT format('%I.%I', ht.schema_name, ht.table_name) AS mat_ht, mat_hypertable_id FROM _timescaledb_catalog.continuous_agg cagg
```

**Review by mkindahl — COMMENTED** [review]
A few nits and some minor comments.

Although I agree that the "correct" solution is to ensure that the invalidation tables are correct rather than force the refresh, I think this option is a useful workaround for those cases where things break.

Although the commit comment is correct and gives a rationale behind why it exists, this is more useful in a design document. I would argue that the commit comment should just contain what the change does and how it works, not why it is implemented.

I also miss a test using an empty table: if you force-refresh an empty range, you could potentially end up with a situation where you try to use a min and max that does not exist, so it's prudent to have a test for that.

File: sql/updates/latest-dev.sql [hunk 4]
```diff
@@ -114,3 +114,14 @@ CREATE FUNCTION @extschema@.hypertable_columnstore_stats (hypertable REGCLASS)
...
+
+-- Recreate `refresh_continuous_aggregate` procedure to add `force` argument
+DROP PROCEDURE IF EXISTS @extschema@.refresh_continuous_aggregate (continuous_aggregate REGCLASS, window_start "any", window_end "any");
+
+CREATE PROCEDURE @extschema@.refresh_continuous_aggregate(
+    continuous_aggregate     REGCLASS,
+    window_start             "any",
+    window_end               "any",
+    force                    BOOLEAN = FALSE
+) LANGUAGE C AS '@MODULE_PATHNAME@', 'ts_update_placeholder';
```
[the remark this answers is not in the document]
Comment (mkindahl) [reply sql/updates/latest-dev.sql:126]: Never mind, this is using the placeholder.

File: tsl/src/continuous_aggs/invalidation.c [hunk 5]
```diff
@@ -140,7 +140,8 @@ static Invalidation cut_cagg_invalidation_and_compute_remainder(
 	const CaggInvalidationState *state, const InternalTimeRange *refresh_window,
 	const Invalidation *mergedentry, const Invalidation *current_remainder);
 static void clear_cagg_invalidations_for_refresh(const CaggInvalidationState *state,
-												 const InternalTimeRange *refresh_window);
+												 const InternalTimeRange *refresh_window,
+												 const bool force);
```
[the remark this answers is not in the document]
Comment (zilder) [reply tsl/src/continuous_aggs/invalidation.c]: agree, I removed `const`s

**Review by mkindahl — APPROVED** [review]

File: tsl/test/expected/cagg_refresh.out (hunk 3 above)
Comment (zilder) [reply tsl/test/expected/cagg_refresh.out]: I changed the test according to your suggestion, though in this particular case this should not be a problem because it is a standard auto-generated name like `_timescaledb_internal._materialized_hypertable_xxx`

Comment (mkindahl) [reply tsl/test/expected/cagg_refresh.out]: Correct, it works now and is not dependent on user data, hence a nit.
